osu.Game.Rulesets.Osu/Scoring/LegacyOsuHealthProcessor.cs

```diff
@@ -0,0 +1,215 @@
+
+using System;
+using System.Linq;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Timing;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Rulesets.Osu.Scoring
+{
+    /// <summary>
+    /// Reference implementation for osu!stable's HP drain.
+    /// Cannot be used for gameplay.
+    /// </summary>
+    public partial class LegacyOsuHealthProcessor : DrainingHealthProcessor
+    {
+        private const double hp_bar_maximum = 200;
+        private const double hp_combo_geki = 14;
+        private const double hp_hit_300 = 6;
+        private const double hp_slider_repeat = 4;
+        private const double hp_slider_tick = 3;
+
+        public Action<string>? OnIterationFail;
+        public Action<string>? OnIterationSuccess;
+        public bool ApplyComboEndBonus { get; set; } = true;
+
+        private double lowestHpEver;
+        private double lowestHpEnd;
+        private double lowestHpComboEnd;
+        private double hpRecoveryAvailable;
+        private double hpMultiplierNormal;
+        private double hpMultiplierComboEnd;
+
+        public LegacyOsuHealthProcessor(double drainStartTime)
+            : base(drainStartTime)
+        {
+        }
+
+        public override void ApplyBeatmap(IBeatmap beatmap)
+        {
+            lowestHpEver = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 195, 160, 60);
+            lowestHpComboEnd = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 198, 170, 80);
+            lowestHpEnd = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 198, 180, 80);
+            hpRecoveryAvailable = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 8, 4, 0);
+
+            base.ApplyBeatmap(beatmap);
+        }
+
+        protected override void ApplyResultInternal(JudgementResult result)
+        {
+            if (!IsSimulating)
+                throw new NotSupportedException("The legacy osu! health processor is not supported for gameplay.");
+        }
+
+        protected override void RevertResultInternal(JudgementResult result)
+        {
+            if (!IsSimulating)
+                throw new NotSupportedException("The legacy osu! health processor is not supported for gameplay.");
+        }
+
+        protected override void Reset(bool storeResults)
+        {
+            hpMultiplierNormal = 1;
+            hpMultiplierComboEnd = 1;
+
+            base.Reset(storeResults);
+        }
+
+        protected override double ComputeDrainRate()
+        {
+            double testDrop = 0.05;
+            double currentHp;
+            double currentHpUncapped;
+
+            do
+            {
+                currentHp = hp_bar_maximum;
+                currentHpUncapped = hp_bar_maximum;
+
+                double lowestHp = currentHp;
+                double lastTime = DrainStartTime;
```

Comment:
Just gonna note that this isn't exactly the same as stable, since `DrainStartTime` comes from `DrainingHealthProcessor` ctor param, which comes from `OsuHealthProcessor` ctor param, which comes from `OsuRuleset.CreateHealthProcessor()` ctor param, which is given as `playableBeatmap.HitObjects[0].StartTime`:

https://github.com/ppy/osu/blob/master/osu.Game/Screens/Play/Player.cs#L230

while stable gives it `hitObjectManager.hitObjects[0].StartTime - hitObjectManager.PreEmpt` (https://github.com/peppy/osu-stable-reference/blob/master/osu!/GameModes/Play/Rulesets/Osu/RulesetOsu_HpCalculation.cs).

Likely not a huge deal in practice, though.

Reply:
@smoogipoo can you confirm you're okay with this?

Reply:
Yes, the value is a best effort but is also accurate to how lazer works - drain starts only from that time.